ipfs-cluster-service/main.go

```diff
@@ -457,69 +457,53 @@ func daemon(c *cli.Context) error {
 
 	// Load all the configurations
 	// always wait for configuration to be saved
-	defer cfg.Shutdown()
+	defer cfgMgr.Shutdown()
 
-	err = cfg.LoadJSONFromFile(configPath)
+	err = cfgMgr.LoadJSONFromFile(configPath)
 	checkErr("loading configuration", err)
 
 	if a := c.String("bootstrap"); a != "" {
-		if len(clusterCfg.Peers) > 0 && !c.Bool("force") {
+		if len(cfgs.clusterCfg.Peers) > 0 && !c.Bool("force") {
 			return errors.New("the configuration provides cluster.Peers. Use -f to ignore and proceed bootstrapping")
 		}
 		joinAddr, err := ma.NewMultiaddr(a)
 		if err != nil {
 			return fmt.Errorf("error parsing multiaddress: %s", err)
 		}
-		clusterCfg.Bootstrap = []ma.Multiaddr{joinAddr}
-		clusterCfg.Peers = []ma.Multiaddr{}
+		cfgs.clusterCfg.Bootstrap = []ma.Multiaddr{joinAddr}
+		cfgs.clusterCfg.Peers = []ma.Multiaddr{}
 	}
 
 	if c.Bool("leave") {
-		clusterCfg.LeaveOnShutdown = true
+		cfgs.clusterCfg.LeaveOnShutdown = true
 	}
 
-	api, err := rest.NewAPI(apiCfg)
-	checkErr("creating REST API component", err)
-
-	proxy, err := ipfshttp.NewConnector(ipfshttpCfg)
-	checkErr("creating IPFS Connector component", err)
-
-	state := mapstate.NewMapState()
-
-	err = validateVersion(clusterCfg, consensusCfg)
-	checkErr("validating version", err)
-
-	tracker := maptracker.NewMapPinTracker(trackerCfg, clusterCfg.ID)
-	mon, err := basic.NewMonitor(monCfg)
-	checkErr("creating Monitor component", err)
-	informer, alloc := setupAllocation(c.GlobalString("alloc"), diskInfCfg, numpinInfCfg)
-
-	cluster, err := ipfscluster.NewCluster(
-		clusterCfg,
-		consensusCfg,
-		api,
-		proxy,
-		state,
-		tracker,
-		mon,
-		alloc,
-		informer)
+	cluster, err := initializeCluster(c, cfgs)
 	checkErr("starting cluster", err)
 
 	signalChan := make(chan os.Signal, 20)
-	signal.Notify(signalChan,
+	signal.Notify(
+		signalChan,
 		syscall.SIGINT,
 		syscall.SIGTERM,
-		syscall.SIGHUP)
+		syscall.SIGHUP,
+	)
+
+	var alreadyExiting bool
 	for {
 		select {
 		case <-signalChan:
-			err = cluster.Shutdown()
-			checkErr("shutting down cluster", err)
+			if alreadyExiting {
```

Comment:
This is potentially racy? If 2 signals are received quickly, it may try to read `alreadyExisting` at the same time it's writing it. Easiest solution is to move `alreadyExiting = true` out of the goroutine.